packages/utils/src/snowflake/bulk-insert-patients.ts

```diff
@@ -2,46 +2,109 @@ import * as dotenv from "dotenv";
 import { v4 as uuidv4 } from "uuid";
 import { getAllPatientIds } from "../patient/get-ids";
+import { elapsedTimeAsStr } from "../shared/duration";
+import { getCxData } from "../shared/get-cx-data";
 
-const sqsClient = new SQSClient({ region: "us-west-2" });
+dayjs.extend(duration);
+
+/**
+ * This script inserts patients into Snowflake, by sending messages to a SQS queue.
+ * The queue is consumed by a Lambda function that generates CSV files and calls a
+ * "transform" lambda.
+ * That second lambda converts the CSV into a diff CSV file and send it to Snowflake.
+ *
+ * Usage:
+ * - set env vars on .env file
+ * - set patientIds array with the patient IDs you want to insert
+ * - run `ts-node src/snowflake/bulk-insert-patients.ts`
+ */
 
 const patientIds: string[] = [];
 
 const cxId = getEnvVarOrFail("CX_ID");
 const apiUrl = getEnvVarOrFail("API_URL");
 const queueUrl = getEnvVarOrFail("FHIR_TO_CSV_QUEUE_URL");
 
+const numberOfParallelExecutions = 20;
+const confirmationTime = dayjs.duration(10, "seconds");
+
+const sqsClient = new SQSClient({ region: getEnvVarOrFail("AWS_REGION") });
 const api = axios.create({ baseURL: apiUrl });
 
 async function main() {
-  const patientsToInsert =
-    patientIds.length > 0 ? patientIds : await getAllPatientIds({ axios: api, cxId });
-  const chunks = chunk(patientsToInsert, 200);
-  for (const chunk of chunks) {
-    await Promise.all(
-      chunk.map(async patientId => {
-        const payload = JSON.stringify({
-          jobId: uuidv4(),
-          cxId,
-          patientId,
-        });
-        sqsClient.sendMessageToQueue(queueUrl, payload, {
+  await sleep(100);
+  const { log } = out("");
+
+  const startedAt = Date.now();
+  log(`>>> Starting...`);
+  const { orgName } = await getCxData(cxId, undefined, false);
+
+  const isAllPatients = patientIds.length < 1;
+  const patientsToInsert = isAllPatients
+    ? await getAllPatientIds({ axios: api, cxId })
+    : patientIds;
+
+  await displayWarningAndConfirmation(patientsToInsert, isAllPatients, orgName, log);
+  log(`>>> Running it...`);
+
+  const failedPatientIds: string[] = [];
+  await executeAsynchronously(
+    patientsToInsert,
+    async patientId => {
+      const payload = JSON.stringify({
+        jobId: uuidv4(),
+        cxId,
+        patientId,
+      });
+      try {
+        await sqsClient.sendMessageToQueue(queueUrl, payload, {
           fifo: true,
           messageDeduplicationId: createUuidFromText(payload),
           messageGroupId: patientId,
         });
-      })
-    );
-    console.log(`Sent ${chunk.length} messages... sleeping for 10 seconds`);
-    await sleep(10000);
+      } catch (error) {
+        log(
+          `Failed to put message on queue for patient ${patientId} - reason: ${errorToString(
+            error
+          )}`
+        );
+        failedPatientIds.push(patientId);
+      }
+    },
+    { numberOfParallelExecutions, minJitterMillis: 100, maxJitterMillis: 200 }
+  );
+
+  log(``);
```

Comment:
intentional?

Reply:
yep, to separate potential errors from the final logs on the output